add work-in-progress overlay

davideg opened the pull request:
Fixes #31 

This adds a one-time overlay explaining that Spaceship Earth is still highly in progress. It sets a cookie to know a user has already seen the overlay and won't show it again for 90 days.

If someone's browser has javascript disabled, no work-in-progress overlay will be displayed which I consider degrading gracefully.

CSS is not my best friend so I'd love any tweaking to make this look nicer and keep the "Check it out" div from floating off the end of the box when the window size shrinks. I'm sure this looks terrible on mobile but haven't set up an emulator to test it 😊

davideg committed:
Add work-in-progress overlay

Adds an overlay with a description explaining that the site is in active
development and solicits feedback.

davideg committed:
remove unnecessary div

davideg committed:
Use cookie to only display wip overlay once

If javascript is enabled, the work-in-progress overlay will be shown once
every 30 days using a cookie to determine if a user has already seen it. If
javascript is disabled, the work-in-progress overlay will not be
displayed.

davideg committed:
ensure cookie name is unique to our app

igor47 commented:
we're using bootstrap in this project, and it sorta seems like you're re-creating their modals: 
https://getbootstrap.com/docs/4.0/components/modal/#live-demo

think you can maybe do this without any css at all?

jeredw commented on spaceship/static/wip-overlay.css:
```diff
@@ -0,0 +1,78 @@
+    transform: translateY(-50%);
+    max-height: 510px;
+    background: rgba(255,255,255, 0.95);
+    h1,h2,h3,p,li,a {
```
vanilla css does not support nested rules (some frameworks like sass do). So you'd have to write this as a separate rule, like .overlay-container h1, .overlay-container h2, ... { (stuff) } below

davideg replied:
Ah, gotcha

jeredw commented on spaceship/static/wip-overlay.css:
```diff
@@ -0,0 +1,78 @@
+      -webkit-font-smoothing: antialiased;
+      -moz-osx-font-smoothing: grayscale;
+    }
+    @media screen and (max-width: $screen-xs-min) {
```
Alas vanilla css doesn't support named constants like this (this looks like it may be another sass thing)? media queries would also have to go at the top level, and then you can specify a selector inside them.

davideg replied:
Very likely the file I was looking at for inspiration must have been for a CSS preprocesser :/

jeredw commented on spaceship/static/wip-overlay.css:
```diff
@@ -0,0 +1,78 @@
+#overlay-background {
```
Personally I avoid id selectors, even when there's only one, cause I've lost way too many specificity wars... :) it's a little obscure but I thought you might be interested because you mentioned css. 
https://dev.to/claireparkerjones/reasons-not-to-use-ids-in-css-4ni4

davideg replied:
Ahhhh...I have a feeling this can explain some crazy CSS behaviors I've pulled my hair out about before.

davideg replied:
![css-specificity-wars](https://user-images.githubusercontent.com/395397/57121565-45972500-6d2d-11e9-89bd-e222c6c76983.png)

jeredw commented on spaceship/static/wip-overlay.css:
```diff
@@ -0,0 +1,78 @@
+    }
+  }
+  #overlay .overlay-title {
+    font-family: 'Circular Std', 'Lato', sans-serif;
```
What is circular std?

davideg replied:
It looked pretty.
https://github.com/elartix/circular-std

But I'm not serving it, so who knows who would actually see it. I'll probably stick to something more standard.

jeredw commented on spaceship/static/wip-overlay.css:
```diff
@@ -0,0 +1,78 @@
+    margin-bottom: 20px;
+  }
+  #overlay .close-button {
+    text-align: center !important;
```
Yep !important is a really common symptom of a specificity issue. :) ("begun, the specificity wars have") Here and below, though, I think it may actually not be necessary.

davideg replied:
👍

jeredw commented on spaceship/static/wip-overlay.js:
```diff
@@ -0,0 +1,45 @@
+    evnt.stopPropagation();
+  };
+  // cookie functions based on https://stackoverflow.com/questions/14573223/set-cookie-and-get-cookie-with-javascript#24103596
+  var setCookie = (name, value, days) => {
```
We have jquery in this project, maybe that has some better utilities to hand already?

davideg replied:
Ah, good to know. I will take a peak at the jQuery way.

jeredw commented:
This is a good call and I'd be happy to help tidy up. Though Igor's got a good point, bootstrap modals may actually just work on mobile too so would be worth looking into

davideg commented:
Thank you @jeredw! I really appreciate your thoughts on the CSS. I found a modal that I liked on the internet and tried to recreate it and didn't know exactly what I was doing 😜.  I agree with @igor47 that if I can do this in bootstrap it would simplify everything. I'll take another stab at this when I get the chance.

davideg committed:
use bootstrap for WIP overlay modal

Rewrites work-in-progress overlay code to use bootstrap with minimal
additional CSS

davideg committed:
change button class to get the pretty hover color

davideg committed:
use js-cookie to set and get cookies for wip modal

Uses cloudflare CDN to source js-cookie. Learn about js-cookie here: https://github.com/js-cookie/js-cookie

davideg committed:
remove unnecessary html class

davideg commented:
I've simplified everything using bootstrap and [js-cookie](https://github.com/js-cookie/js-cookie). Let me know what you think.

Here's what it looks like now:
![image](https://user-images.githubusercontent.com/395397/57065685-11fdc180-6c7f-11e9-9a48-292a762e1775.png)

jeredw commented on spaceship/static/wip-overlay.js:
```diff
@@ -0,0 +1,10 @@
+(function() {
+  window.addEventListener('load', () => {
```
$(window).load(function() { ... }); is idiomatic for jquery.

davideg replied:
Looks like $.load is deprecated, but I've changed it to use $.on

jeredw commented:
Nice, thanks! I tried it out in mobile emulation in Chrome, and looks great. I also checked that the cookie stuff works properly with the existing session cookie.

davideg committed:
use jquery to load wip overlay
